### ISSUE TITLE
Campaigns default reviewers in bitbucket server

### ISSUE THREAD

**Issue by jensim** [sourcegraph/sourcegraph#14746]
#### Feature request description

Add default reviewers to PRs made when applying campaigns

#### Is your feature request related to a problem? If so, please describe.

In bitbucket server dashboard users/codeowners are only notified of PRs that they are added to. 
And sourcegraph campaigns do not add these people to the views, leaving a massive amount of manual labour (or scripting), that we want to avoid.

#### Describe alternatives you've considered.

Use bitbucket server apis for getting the suggested reviewers given 
- sourceRepoID
- targetRepoID
- sourceRef
- targetRef
```
curl -u "${username}:${password}" "https://my-bitbucket-server.local/rest/default-reviewers/1.0/projects/$key/repos/$repo_slug/reviewers?sourceRefId=${my_branch}&targetRefId=master&sourceRepoId=${repo_id}&targetRepoId=${repo_id}"
```

**Comment by jensim** [comment]
I've now created a bash script for adding default reviewers to all PRs that the executing user is author to, and have zero reviewers.

https://gist.github.com/jensim/c157b935f3c3874bd9857509a6b36f82

The docs for bitbucket server rest api are here  
https://docs.atlassian.com/bitbucket-server/rest/5.16.0/bitbucket-rest.html#idm8287078800

**Comment by chrispine** [comment]
Thanks for reaching out, @jensim! This is great feedback. We have added this to our roadmap, and will work to find a time to slot this in.

**Comment by chrispine** [comment]
@eseliger Could you please add an estimate to this? Thanks!

**Comment by eseliger** [comment]
I think the tricky part here is identifying when we want to do this. Adding a second REST call to the `CreateChangeset` method of the BitbucketServerSource shouldn't be super hard.
But the open question would be whether we _always_ want to do this, add a bitbucket specific setting to the campaign spec, ... [1]
If we decide on something, I don't think it'll be more than a days work, given it's adding a simple rest call and some testing/documentation work.

---
[1] Other funky ideas here.

**Comment by mrnugget** [comment]
It's two rest calls, right? We need to get the list of suggested reviewers and then we need to update the pull request to add those reviewers.

I think if we do this, we should do it properly instead of just adding two additional calls to the Bitbucket Server client. 

Why? Because
- GitHub also has suggested reviewers — should we do it for GitHub too? If so, not in draft mode, I assume.
- Wouldn't it be better to have a `reviewers:` field in the `changesetTemplate`? Default is an empty list, can be a list of usernames, or a globbing-pattern+list-of-usernames, or "suggested".

Then you could do:

```yaml
changesetTemplate:
  # [...]
  published: true
  reviewers: suggested
```

or

```yaml
changesetTemplate:
  # [...]
  published: true
  reviewers:
    - *: []
    - github.com/sourcegraph/*: [@eseliger]
    - github.com/src-cli/*: suggested
```

---

In any case: I'm going to remove the 1d estimate because I think it's too low and I want to discuss this.

**Comment by jensim** [comment]
Really nice suggestions @mrnugget , I think the second one looks like a good format. Having multiple formats might be confusing, don't you think?
Like i mentioned in my previous post, I've got a working script for adding default reviewers on Bitbucket server, so I'm covered there, and in no rush 😃

**Comment by mrnugget** [comment]
> Having multiple formats might be confusing, don't you think?

Sorry, I wasn't super clear there and the "or" makes it sound like it's two options. But it's actually the same format: you can specify `reviewers` for all changesets or for a subset. We already allow the same behaviour with the `published` field, see [the examples here](https://docs.sourcegraph.com/campaigns/references/campaign_spec_yaml_reference#changesettemplate-published).

**Comment by eseliger** [comment]
@jensim this has been implemented and will be part of Sourcegraph 3.28 as the default behavior for Bitbucket Server PRs. Thanks again for reporting!
